### PR TITLE
ui: Issues page

### PR DESCRIPTION
**Pull request by Etsija**
Please see the commits for details.

An open question: I believe the issues are sorted by timestamp by default. It could make the issues table cleaner if the back-end sorting would be hard-coded in the UI into severity instead. Another possibility would be to do also sorting on client side, but this hasn't yet been implemented in this PR.

### CODE REVIEW

File: ui/src/helpers/get-status-class.ts [hunk 1]
```diff
@@ -17,27 +17,23 @@
...
   | 'HIGH'
   | 'MEDIUM'
   | 'LOW'
   | 'NONE';
 
-type RuleViolationSeverity = 'ERROR' | 'WARNING' | 'HINT';
+// In case these differ in the future in OpenApi spec, keep them as separate types here,
+// but they could also very well be combined for simplicity.
+type RuleViolationSeverity = Severity;
+type IssueSeverity = Severity;
```
Comment (mmurto) [inline ui/src/helpers/get-status-class.ts:36]: I'd combine these as there shouldn't be any plans to separate them.

File: ui/src/schemas/index.ts [hunk 2]
```diff
@@ -43,3 +43,15 @@ export const ortRunStatus: z.ZodEnum<[OrtRunStatus, ...OrtRunStatus[]]> =
...
+export const issueSeverity: z.ZodEnum<[Severity, ...Severity[]]> = z.enum([
+  'HINT',
+  'WARNING',
+  'ERROR',
+]);
+
+// Issue severity schema that is used for search parameter validation
+export const issueSeveritySchema = z.object({
+  severity: z.array(issueSeverity).optional(),
+});
```
Comment (mmurto) [inline ui/src/schemas/index.ts:57]: More idiomatic zod naming would be that `issueSeverity` is `issueSeveritySchema` and the current `issueSeveritySchema` something different.

File: ui/src/test-data/mock-up-issues.tsx [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+ */
+
+// Some mock-up data to test the functionality of the Issues data table.
+// Data was created with ChatGPT, and for most part, includes made up
+// random data, not reflecting the real data that would be returned from
+// an ORT run.
+
+import { Severity } from '@/api/requests';
+
+export const mockUpIssues = {
```
Comment (mmurto) [inline ui/src/test-data/mock-up-issues.tsx:27]: As we don't have this sort of data anywhere else, I wouldn't commit it here either.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/issues/index.tsx [hunk 4]
```diff
@@ -18,67 +18,282 @@
...
+  row: Row<Issue>,
+  columnId: string,
+  filterValue: string[]
+): boolean => {
+  // TypeScript can't infer the type, so cast it explicitly.
+  // This is safe because the columnId is always one of the
+  // severities, which themselves are strings.
+  const rowValue = row.getValue(columnId) as string;
+  return filterValue.includes(rowValue);
+};
```
Comment (mmurto) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/issues/index.tsx:81]: As this is only used in one column, I would move the declaration in-line to the column definition, mostly in hope to gain better type-safety.

**Review by mmurto — CHANGES_REQUESTED** [review]

**Comment by mmurto** [comment]
> Another possibility would be to do also sorting on client side, but this hasn't yet been implemented in this PR.

Doesn't this already do sorting client-side, or did I misread?

**Comment by Etsija** [comment]
> > Another possibility would be to do also sorting on client side, but this hasn't yet been implemented in this PR.
> 
> Doesn't this already do sorting client-side, or did I misread?

No sorting on client-side is implemented.

**Comment by mmurto** [comment]
> > > Another possibility would be to do also sorting on client side, but this hasn't yet been implemented in this PR.
> > 
> > 
> > Doesn't this already do sorting client-side, or did I misread?
> 
> No sorting on client-side is implemented.

Ah yeah, sorting is not implemented at all, so not server-side either.

File: ui/src/schemas/index.ts (hunk 2 above)
Comment (Etsija) [reply ui/src/schemas/index.ts:57]: I mimicked the naming from Johanna's definitions earlier in that file, so they are in line with each other.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/issues/index.tsx (hunk 4 above)
Comment (Etsija) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/issues/index.tsx]: It didn't, though.

Comment (mmurto) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/issues/index.tsx]: Looking at it here, does it just filter the exact values? Would some of the built-in functions work like:

```typescript
    filterFn: 'equalsString',
```

Comment (Etsija) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/issues/index.tsx]: > Looking at it here, does it just filter the exact values? Would some of the built-in functions work like:
> 
> ```typescript
>     filterFn: 'equalsString',
> ``` 

No, it filters against an array of possible severities, like `[WARNING, ERROR]` and there are no built-in functions for that. There are three functions that do the exact opposite - filter an array of values against an option - but no built-in functions for filtering a value against an array of options. All the solutions I could find used `includes()`.

Comment (mmurto) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/issues/index.tsx]: Thanks, makes sense.

Comment (mmurto) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/issues/index.tsx]: You can make it more type-safe like this:

```typescript
    filterFn: (
      row,
      _columnId,
      filterValue
    ): boolean => {
      return filterValue.includes(row.original.severity);
    },
```

Comment (Etsija) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/issues/index.tsx]: That was a nice simplifying fix, thank you!

**Review by mmurto — APPROVED** [review]

**Pull request merged**